### PR TITLE
[Feature](ElasticEP) Implement cleanup and retry logic to recover from transient error

### PR DESCRIPTION
**Pull request by fangyuchu**
## Purpose
This PR implements a fault tolerance mechanism for transient and recoverable failures, based on the pause-on-error capability introduced in https://github.com/vllm-project/vllm/pull/38534. It enables the system to pause on failure, clean up request states, and resume execution via retry without user-visible errors.

This PR is the third part of the split plan from the initial PR https://github.com/vllm-project/vllm/pull/28296.

## Mechanism

Based on https://github.com/vllm-project/vllm/pull/38534:

- When a failure occurs, the buffer mask maintained by `All2AllManager` automatically marks the failed rank.
- For transient and recoverable failures (e.g., network glitches), the serving framework can issue a retry command via `/fault_tolerance/apply`.

During retry:

- The sentinel refreshes the engine state.
- The `cpu_group` in the DP group is rebuilt (as it may have timed out during failure).
- The rank mask is reset to all zeros.

Request handling:

- All in-flight requests are preempted and moved back to the waiting queue.
- The step counter in the engine core is reset.
- After retry, the engine reschedules requests from the waiting queue.

On retry, we rely on prefix matching to reuse KV cache as KV blocks are only released logically (i.e., we free their usage indices), but the underlying data remains intact. As a result, previously computed tokens can be reused. Only the tail tokens (e.g., those not filling a full KV block) need to be recomputed, similar to a chunked prefill.

From the user’s perspective:

- The inference may pause briefly.
- No request failures or duplicated tokens are observed.

Note:
- Worker-side states (e.g., `input_batch`) are cleared during retry.

### Unit Tests

```bash
pytest tests/v1/fault_tolerance/ -v -s
```
### End-to-End Test

1. Start the service:
    
    ```bash
    vllm serve \
      --model xxx \
      --tensor-parallel-size 1 \
      --data-parallel-size 4 \
      --port 8119 \
      --all2all-backend deepep_low_latency \
      --enable_expert_parallel \
      --enable-fault-tolerance \
      --gloo-timeout-seconds 120 \
      --fault-tolerance-config='{"engine_recovery_timeout_sec":120,"internal_fault_report_port":xxxx,"external_fault_notify_port":xxxx}'
    ```

2. Inject a failure in a specific rank at a specific step (similar to #35193).

3. Wait for vLLM to pause.

4. Send retry request via `/fault_tolerance/apply`.
    ```bash
    curl -X POST "http://x.x.x.x:8119/fault_tolerance/apply" \
      -H "Content-Type: application/json" \
      -d '{"instruction":"retry","params":{"timeout":10}}'
    ```

## Test Results

- **Unit tests**: Passed.

- **End-to-end**:
    - After fault injection and waiting for DeepEP kernel timeout (~100s), logs show:
      ```
      vLLM instance is paused and waiting for recovery commands.
      ```
    - After sending retry:
        - Recovery command succeeds.
        - Previously stalled requests resume and continue returning outputs.

---

## Acknowledgments

We would like to thank the following contributors for their valuable contributions:

- @zWaNg3
- @a798347923
- @205150940
- @yzchang-plus
- @jianzs

We also appreciate the community for the numerous help and suggestions on the fault tolerance framework.

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR.
- [x] The test plan.
- [x] The test results.
- [ ] (Optional) Documentation updates.

</details>

### CODE REVIEW

**Comment by tzulingk** [comment]
The architectural alternative I'd propose for this stack is on #34833: https://github.com/vllm-project/vllm/pull/34833#issuecomment-4349331831
One bit of that conversation is relevant to **this** PR specifically: `gemini-code-assist`'s "worker results dropped" inline comments on `EngineCoreSentinel.pause` / `retry`, plus the `if True` dead-code in `retry`'s return, are direct symptoms of the new ZMQ DEALER/ROUTER channel re-implementing what `Executor.collective_rpc` already provides. With `collective_rpc`, the natural calling pattern is `results = collective_rpc(...)` and the result can't be silently dropped. The bug class disappears by construction. Detail and pseudocode at the link above.
Independently of that architectural discussion, two specific cleanups on FT mask handling — see inline comments below.

Not asking for rework before #34833 settles. The retry semantics in this PR (preempt-to-waiting + prefix-cache reuse) are correct in spirit; the wrapping infrastructure is what the architectural conversation is about.

File: vllm/distributed/device_communicators/all2all.py
Comment (tzulingk) [inline vllm/distributed/device_communicators/all2all.py:1]: ### Move `query_mask` / `clean_mask` to a capability mixin

These methods aren't specific to the two FT all2all backends in this PR (`DeepEPLLAll2AllManager`, `NixlEPAll2AllManager`) — they describe a **capability** that any FT-capable collective backend can implement. **FT NCCL collectives** (cf. `vllm/distributed/eplb/ft_nccl_wrapper.py` work-in-progress) are the natural near-term addition; future FT-capable backends will be too.

The right surface for that capability is a mixin, not the abstract base. As written, every `BaseDeviceCommunicator` subclass — `PyNcclCommunicator`, `CustomAllReduce`, `XpuCommunicator`, the CPU communicator, etc. — inherits methods that raise `NotImplementedError` for a capability that doesn't apply to them. That's both dead surface area for non-FT backends and a missing forward-compat story for new FT backends (each one has to silently override the same two methods, with no way for callers to ask "does this thing support FT masking?").

#### Suggested change

Define a capability mixin:

```python
# vllm/v1/fault_tolerance/maskable.py — new (or in fault_tolerance/__init__.py)
from typing import Protocol, runtime_checkable
import torch

@runtime_checkable
class FTMaskBuffer(Protocol):
    """Capability mixin for FT-capable collective backends.

    The kernel auto-sets a bit for any peer it deems faulty (e.g., DeepEP-LL
    after timeout, NIXL-EP after timeout, FT NCCL when the collective fails).
    `query_mask` reads which peers are flagged; `clean_mask` resets to zero.
    See vLLM FT design doc §7.2 for the lifecycle.

    Implemented today by:
      - DeepEPLLAll2AllManager
      - NixlEPAll2AllManager
    Likely future implementers:
      - FT NCCL collectives (cf. vllm/distributed/eplb/ft_nccl_wrapper.py)
      - any other backend that exposes peer-failure masking
    """
    def query_mask(self, mask: torch.Tensor) -> torch.Tensor: ...
    def clean_mask(self) -> None: ...
```

Attach to backends that actually implement it:

```python
# vllm/distributed/device_communicators/all2all.py
from vllm.v1.fault_tolerance.maskable import FTMaskBuffer

class DeepEPLLAll2AllManager(FTMaskBuffer, DeepEPAll2AllManagerBase):
    # existing query_mask / clean_mask methods stay HERE
    ...

class NixlEPAll2AllManager(FTMaskBuffer, All2AllManagerBase):
    # existing query_mask / clean_mask methods stay HERE
    ...
```

Or for the FT NCCL case, on the appropriate level (e.g., the device communicator itself, or a wrapper that owns the FT NCCL handle):

```python
# Future / WIP: vllm/distributed/eplb/ft_nccl_wrapper.py
class FTNCCLWrapper(FTMaskBuffer):
    def query_mask(self, mask): ...
    def clean_mask(self): ...
```

Remove the abstract stubs from `BaseDeviceCommunicator`:

```python
# vllm/distributed/device_communicators/base_device_communicator.py — REVERT
class BaseDeviceCommunicator:
    def dispatch(self, ...) -> ...: ...
    # (no query_mask / clean_mask here — capability lives on FTMaskBuffer)
```

Callers ask whether the capability is present, instead of catching `NotImplementedError`:

```python
# Anywhere in the FT framework that needs to confirm a peer
candidates = [
    comm.all2all_manager,                 # FT all2all backends
    getattr(comm, "ft_nccl", None),       # FT NCCL wrapper, if present
]
for x in candidates:
    if isinstance(x, FTMaskBuffer):
        mask = x.query_mask(mask_buffer)
        # ... use mask ...
        break
```

#### Why this is the right shape

1. **Capability-first, backend-agnostic.** The mixin says "this object can be queried for FT mask state," without committing to which kind of backend. DeepEP-LL, NIXL-EP, FT NCCL, and anything FT-capable in the future all match the same contract.

2. **No dead surface area on non-FT backends.** `PyNcclCommunicator`, `CustomAllReduce`, the CPU communicator etc. don't carry FT methods that they can never implement.

3. **The capability is discoverable.** `isinstance(x, FTMaskBuffer)` is a positive test. Today the only way to know is to call the method and see if it raises.

4. **Forward-compatible with FT NCCL work.** When `vllm/distributed/eplb/ft_nccl_wrapper.py` lands, it implements the same mixin and immediately works with the same supervisor / cleanup paths. No further edits to `BaseDeviceCommunicator`, no new branches in callers.

This is consistent with vLLM's existing capability mixins (`SupportsLoRA`, `SupportsPP`, etc.) — that's how vLLM expresses optional capabilities elsewhere in the codebase.

File: vllm/v1/worker/sentinel/gpu_worker_sentinel.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            self.engine_core_cmd_socket.send_multipart([b"", msg_bytes])
+        except zmq.ZMQError:
+            logger.info("Socket closed, terminating.")
+            self.sentinel_dead = True
+
+    def pause(self, ft_request: FaultToleranceRequest) -> FaultToleranceResult:
+        get_pause_event().set()
+        return FaultToleranceResult(ft_request.request_id, True)
+
+    def retry(self, ft_request: FaultToleranceRequest) -> FaultToleranceResult:
```
Comment (tzulingk) [inline vllm/v1/worker/sentinel/gpu_worker_sentinel.py:104]: Can you add comments what retry is trying to do?
It's not retrying the failed request, it's clearing up states.

File: vllm/v1/fault_tolerance/engine_core_sentinel.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        Handle the retry instruction from the ClientSentinel.
+        This instruction tells the EngineCore to continue its busy loop
+        after being suspended due to an exception.
+        """
+        if not self.busy_loop_paused.is_set():
+            return FaultToleranceResult(ft_request.request_id, True)
+        timeout = ft_request.params["timeout"]
+        self.parallel_config._coord_store_port = ft_request.params["coord_store_port"]
+        self._execute_command_on_workers(
+            FaultToleranceRequest(str(uuid.uuid4()), "retry", ft_request.params),
```
Comment (tzulingk) [inline vllm/v1/fault_tolerance/engine_core_sentinel.py]: why do we want to regenerate the request id here? Why not use the original one?
anyone tracing logs by request_id won't be able to follow the call into the workers.

File: vllm/v1/fault_tolerance/engine_core_sentinel.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+            self.worker_identities,
+            timeout=timeout,
+        )
+        if self.data_parallel_size > 1:
+            # If the Gloo communication times out,
+            # the data parallel group (dp_group) needs to be reinitialized
+            reinit_request = FaultToleranceRequest(
+                instruction="reinit_dp_group_on_fault_tolerance",
+                request_id=str(uuid.uuid4()),
```
Comment (tzulingk) [inline vllm/v1/fault_tolerance/engine_core_sentinel.py]: same here, why generate new id?

File: vllm/v1/fault_tolerance/engine_core_sentinel.py (hunk 2 above)
Comment (fangyuchu) [reply vllm/v1/fault_tolerance/engine_core_sentinel.py]: Fixed as suggested.

**Comment by tzulingk** [comment]
The new structure already distinguishes "instruction arrived → `continue`" from "timed out → log error → `raise`". But `engine_recovery_timeout_sec` only makes sense when an external supervisor is actually expected to send an instruction. `report_fault_events()` clears `run_busy_loop` before the wrapper calls `run_busy_loop.wait(timeout=...)`, so in the default (no-orchestrator) deployment the wrapper blocks for the full timeout before the inevitable `raise` — pure dead time.

Suggest splitting into two modes:

- **`"degrade"` (default, no supervisor):** branch *before* the `wait()`. Log a warning (`"EngineCore fault: no supervisor configured; continuing in degraded mode"`), re-set `run_busy_loop`, `continue`.
- **`"supervised"`:** current behavior — `wait(timeout=engine_recovery_timeout_sec)` → `continue` on recovery or `raise` on timeout.

```python
mode = self.vllm_config.fault_tolerance_config.on_fault  # "degrade" | "supervised"

if mode == "degrade":
    logger.warning(
        "[BusyLoopWrapper] EngineCore fault: no supervisor configured; "
        "continuing in degraded mode."
    )
    self.sentinel.run_busy_loop.set()
    continue

recovered = self.sentinel.run_busy_loop.wait(
    timeout=self.engine_recovery_timeout_sec
)
if recovered:
    continue
logger.error("[BusyLoopWrapper] EngineCore did not recover in time.")
raise
```

Also tighten the `engine_recovery_timeout_sec` docstring so it no longer asserts "the original error is raised" as the only outcome — it's a supervisor-mode-only knob.
